src/picongpu/include/particles/bremsstrahlung/PhotonEmissionAngle.hpp

```diff
@@ -0,0 +1,264 @@
+/**
+ * Copyright 2016 Heiko Burau
+ *
+ * This file is part of PIConGPU.
+ *
+ * PIConGPU is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PIConGPU is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with PIConGPU.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "cuSTL/container/HostBuffer.hpp"
+#include "cuSTL/cursor/Cursor.hpp"
+#include "cuSTL/cursor/navigator/PlusNavigator.hpp"
+#include "cuSTL/cursor/tools/LinearInterp.hpp"
+#include "cuSTL/cursor/BufferCursor.hpp"
+#include "algorithms/math.hpp"
+#include <boost/array.hpp>
+#include <boost/shared_ptr.hpp>
+#include <boost/numeric/odeint/integrate/integrate.hpp>
+#include <boost/math/tools/minima.hpp>
+#include <limits>
+#include <utility>
+
+namespace picongpu
+{
+namespace particles
+{
+namespace bremsstrahlung
+{
+
+namespace detail
+{
+
+/** Functor mapping `delta` to the photon emission polar angle `theta`,
+ * where delta is a uniformly distributed random number between zero and one.
+ */
+struct GetPhotonAngleFunctor
+{
+    typedef typename ::PMacc::result_of::Functor<
+        ::PMacc::cursor::tools::LinearInterp<float_X>,
+        ::PMacc::cursor::BufferCursor<float_X, DIM2> >::type LinInterpCursor;
+
+    typedef float_X type;
+
+    LinInterpCursor linInterpCursor;
+    float_X lnMinGamma;
+    float_X lnMaxGamma;
+
+    /** constructor
+     *
+     * @param linInterpCursor lookup table for the photon emission angle.
+     */
+    HDINLINE GetPhotonAngleFunctor(LinInterpCursor linInterpCursor)
+        : linInterpCursor(linInterpCursor)
+    {
+        this->lnMinGamma = algorithms::math::log(photon::MIN_GAMMA);
+        this->lnMaxGamma = algorithms::math::log(photon::MAX_GAMMA);
+    }
+
+    /** Return the polar emission angle of the photon.
+     *
+     * @param delta uniformly distributed random number between zero and one.
+     * @param gamma relativistic factor of the incident electron.
+     */
+    HDINLINE float_X operator()(const float_X delta, const float_X gamma) const
+    {
+        const float_X binDelta = delta * static_cast<float_64>(photon::NUM_SAMPLES_DELTA - 1);
+
+        const float_X lnGamma = algorithms::math::log(gamma);
+        const float_X binGamma =
+            (lnGamma - this->lnMinGamma) /
+            (this->lnMaxGamma - this->lnMinGamma) *
+            static_cast<float_X>(photon::NUM_SAMPLES_GAMMA - 1);
+
+        if (picLog::log_level & picLog::CRITICAL::lvl)
+        {
+            if(binGamma > float_X(photon::NUM_SAMPLES_GAMMA - 1))
```

Comment:
IMO it should by `if(binGamma >= photon::NUM_SAMPLES_GAMMA )`
`binGamma` should be an integer type because there are no half bins. This also avoid aut of memory access.